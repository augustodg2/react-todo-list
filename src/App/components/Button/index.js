import React from 'react'
import PropTypes from 'prop-types'

const Button = ({ variant, action, icon, text }) => {
  const hasVariant = (wantedVariant, variants) => {
    return variant.split(' ').includes(wantedVariant)
  }

  return (
    <button
      data-testid="button"
      className={ `button ${variant}` }
      onClick={ action }
      type={ hasVariant('primary') ? 'submit' : 'button' }
    >
      { icon }
      <span data-testid="button__text">{ text }</span>
    </button>
  )
}

Button.defaultProps = {
  variant: ''
}

Button.propTypes = {
  variant: PropTypes.string,
  action: PropTypes.func,
  icon: PropTypes.element,
  text: PropTypes.string
}

export default Button
