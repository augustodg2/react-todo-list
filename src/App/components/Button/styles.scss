.button {
  display: flex;
  background: none;
  outline: none;
  border: none;
  color: $dark-gray;
  font-family: 'Open Sans';
  font-size: 0.875rem;
  cursor: pointer;
  
  &.primary {
    background: $primary;
    color: $white;
    border: none;
    padding: .5rem 1rem;
    border-radius: .3rem;
  }

  &.secondary {
    background: $light-gray;
    color: $black;
    border: none;
    padding: .5rem 1rem;
    border-radius: .3rem;

    & svg {
      color: $dark-gray
    }
  }

  &.danger {
    background: $red;
    color: $white;
    border: none;
    padding: .5rem 1rem;
    border-radius: .3rem;
  }

  &.round {
    border-radius: 3rem;
    padding: .5rem;
  }
  
  &.float {
    padding: 1rem;
    box-shadow: 0 0 2rem rgba($color: #000000, $alpha: 0.4);
  }
}