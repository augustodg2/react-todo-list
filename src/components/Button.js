import React from 'react'
import PropTypes from 'prop-types'

const Button = ({ variant, action, icon, text }) => {
  const hasVariant = (wantedVariant, variants) => {
    return variant.split(' ').includes(wantedVariant)
  }

  return (
    <button
      className={ `button ${variant}` }
      onClick={ action }
      type={ hasVariant('primary') ? 'submit' : 'button' }
    >
      <i className='material-icons'>{ icon }</i>
      <span>{ text }</span>
    </button>
  )
}

Button.propTypes = {
  variant: PropTypes.string.isRequired,
  action: PropTypes.func,
  icon: PropTypes.string,
  text: PropTypes.string
}

export default Button
