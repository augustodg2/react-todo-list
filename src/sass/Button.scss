.button {
  background: none;
  outline: none;
  border: none;
  color: $dark-gray;
  font-family: 'Open Sans';
  font-size: 0.875rem;
  
  &.primary {
    background: $primary;
    color: $white;
    border: none;
    padding: .5rem 1rem;
    border-radius: .3rem;
  }

  &.secondary {
    background: $gray;
    color: $black;
    border: none;
    padding: .5rem 1rem;
    border-radius: .3rem;
  }

  &.round {
    height: 3rem;
    width: 3rem;
    border-radius: 3rem;

    i {
      font-size: 1rem;
    }
  }
}