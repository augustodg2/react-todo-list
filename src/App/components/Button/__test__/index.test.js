import React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import Button from '../index'

afterEach(cleanup)

describe('# Button', () => {
  // Snapshot
  it('renders correctly', () => {
    const { getByTestId } = render(
      <Button text="click me" icon="add" />
    )
    const button = getByTestId('button')
    expect(button).toMatchSnapshot()
  })

  // Button text
  it('should render text from props', () => {
    const { getByTestId } = render(
      <Button text="click me" />
    )
    const buttonText = getByTestId('button__text')
    expect(buttonText.textContent).toBe('click me')
  })

  // Button icon
  it('should render icon from props', () => {
    const { getByTestId } = render(
      <Button text="click me" icon="add" />
    )
    const buttonIcon = getByTestId('button__icon')
    expect(buttonIcon.textContent).toBe('add')
  })

  // Button variants
  it('should add variant classnames to button', () => {
    const { getByTestId } = render(
      <Button variant="abc abcd" />
    )
    const button = getByTestId('button')
    expect(button.className).toContain('abc abcd')
  })

  // Action
  it('should execute action on click', () => {
    const action = jest.fn()
    const { getByTestId } = render(
      <Button action={action} />
    )

    const button = getByTestId('button')
    fireEvent.click(button)

    expect(action).toBeCalledTimes(1)
  })
})
