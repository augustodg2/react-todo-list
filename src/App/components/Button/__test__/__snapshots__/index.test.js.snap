// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# Button renders correctly 1`] = `
<button
  class="button "
  data-testid="button"
  type="button"
>
  <i
    class="material-icons"
    data-testid="button__icon"
  >
    add
  </i>
  <span
    data-testid="button__text"
  >
    click me
  </span>
</button>
`;
